lib/amplitude/plugin.rb

```diff
@@ -0,0 +1,132 @@
+require 'securerandom'
+module AmplitudeAnalytics
+  # Plugin
+  class Plugin
+    attr_reader :plugin_type
+
+    def initialize(plugin_type)
+      @plugin_type = plugin_type
+    end
+
+    def setup(client)
+      # Setup plugins with client instance parameter
+    end
+
+    def execute(event)
+      # Process event with plugin instance
+    end
+  end
+
+  # EventPlugin
+  class EventPlugin < Plugin
+    def execute(event)
+      track(event)
+    end
+
+    def track(event)
+      event
+    end
+  end
+
+  # DestinationPlugin
+  class DestinationPlugin < EventPlugin
+    attr_reader :timeline
+
+    def initialize
+      super(PluginType::DESTINATION)
+      @timeline = Timeline.new
+    end
+
+    def setup(client)
+      @timeline.setup(client)
+    end
+
+    def add(plugin)
```

Comment:
Do we need `add` and `remove` in Destination Plugin itself?

Reply:
Mirroring [Python SDK](https://github.com/amplitude/Amplitude-Python/blob/main/src/amplitude/plugin.py#L95), this is so that we can "add additional processing ability by plugins to modify events before sending out"